auth/integration_test/src/integration_test.cc

```diff
@@ -860,57 +873,67 @@ TEST_F(FirebaseAuthTest, TestPhoneAuth) {
   // Note: This test requires interactivity on iOS, as it displays a CAPTCHA.
   TEST_REQUIRES_USER_INTERACTION;
 #endif  // TARGET_OS_IPHONE
-  {
-    firebase::auth::PhoneAuthProvider& phone_provider =
-        firebase::auth::PhoneAuthProvider::GetInstance(auth_);
-    LogDebug("Creating listener.");
-    PhoneListener listener;
-    LogDebug("Calling VerifyPhoneNumber.");
-    // Randomly choose one of the phone numbers to avoid collisions.
-    const int random_phone_number =
-        app_framework::GetCurrentTimeInMicroseconds() %
-        kPhoneAuthTestNumPhoneNumbers;
-    phone_provider.VerifyPhoneNumber(
-        kPhoneAuthTestPhoneNumbers[random_phone_number], kPhoneAuthTimeoutMs,
-        nullptr, &listener);
-    // Wait for OnCodeSent() callback.
-    int wait_ms = 0;
-    LogDebug("Waiting for code send.");
-    while (listener.waiting_to_send_code()) {
-      if (wait_ms > kPhoneAuthCodeSendWaitMs) break;
-      ProcessEvents(kWaitIntervalMs);
-      wait_ms += kWaitIntervalMs;
-    }
-    EXPECT_EQ(listener.on_verification_failed_count(), 0);
-    LogDebug("Waiting for verification ID.");
-    // Wait for the listener to have a verification ID.
-    wait_ms = 0;
-    while (listener.waiting_for_verification_id()) {
-      if (wait_ms > kPhoneAuthCompletionWaitMs) break;
-      ProcessEvents(kWaitIntervalMs);
-      wait_ms += kWaitIntervalMs;
-    }
-    if (listener.on_verification_complete_count() > 0) {
-      LogDebug("Signing in with automatic verification code.");
-      WaitForCompletion(auth_->SignInWithCredential(listener.credential()),
-                        "SignInWithCredential(PhoneCredential) automatic");
-    } else if (listener.on_verification_failed_count() > 0) {
-      FAIL() << "Automatic verification failed.";
-    } else {
-      // Did not automatically verify, submit verification code manually.
-      EXPECT_GT(listener.on_code_auto_retrieval_time_out_count(), 0);
-      EXPECT_NE(listener.verification_id(), "");
-      LogDebug("Signing in with verification code.");
-      const firebase::auth::Credential phone_credential =
-          phone_provider.GetCredential(listener.verification_id().c_str(),
-                                       kPhoneAuthTestVerificationCode);
-
-      WaitForCompletion(auth_->SignInWithCredential(phone_credential),
-                        "SignInWithCredential(PhoneCredential)");
-    }
+  if (!RunFlakyBlock([&]() {
```

Comment:
If you wanted to be really fancy, I bet you could create a FLAKEY_TEST_F macro, that was defined as something like:

    enum class FlakeResult { FLAKEY_FAILURE, FLAKEY_SUCCESS };

    #define FLAKEY_TEST_F(TestHarness, TestName)                      \
                                                                      \
    class TestHarness ## _ ## TestName ## _ ## FlakeHarness {         \
      FlakeResult RunTest();                                          \
    }                                                                 \
                                                                      \
    TEST_F(TestHarness, TestName) {                                   \
      if (FLAKEY_FAILURE == RunFlakyBlock([&]() {                     \
        TestHarness ## _ ## TestName ## _ ## FlakeHarness().RunTest() \
      })) {                                                           \
        FAIL() << "Your flakey test failed.";                         \
      }                                                               \
    }                                                                 \
                                                                      \
    FlakeResult TestHarness ## _ ## TestName ## _ ## FlakeHarness::RunTest()

(This is just a proof of concept, and I haven't tested or even tried to compile it - but I think something of this general form would work)

Then you could just write the test like this:

    FLAKEY_TEST_F(FirebaseAuthTest, TestPhoneAuth) {
      // Do the test
      return FLAKEY_SUCCESS;
    }

If there is work you only want to do once, you could throw it in the Setup and Teardown functions, and It hink that would work for you.

Reply:
This is a very interesting idea - I'm going to attempt this in a subsequent PR.

Reply:
sgtm